src/eos-updater-poll-common.c

```diff
@@ -386,41 +391,37 @@ get_refspec_to_upgrade_on (gchar               **refspec_to_upgrade_on,
   if (booted_deployment == NULL)
     return FALSE;
 
-  if (!get_refspec_to_upgrade_on_from_deployment (sysroot,
-                                                  booted_deployment,
-                                                  &refspec,
-                                                  error))
+  if (!get_booted_refspec (booted_deployment,
+                           &refspec,
+                           &remote,
+                           &ref,
+                           &collection_ref,
+                           error))
     return FALSE;
 
-  /* Nothing here */
-  if (refspec == NULL)
-    return get_booted_refspec (booted_deployment,
-                               refspec_to_upgrade_on,
-                               remote_to_upgrade_on,
-                               ref_to_upgrade_on,
-                               collection_ref_to_upgrade_on,
-                               error);
-
-  /* We'll need the repo to get the collection-id */
-  if (!ostree_sysroot_get_repo (sysroot, &repo, NULL, error))
+  if (!get_ref_to_upgrade_on_from_deployment (sysroot,
+                                              booted_deployment,
+                                              &ref_for_deployment,
+                                              error))
     return FALSE;
 
-  /* Found a refspec on this commit's metadata */
-  if (!ostree_parse_refspec (refspec, &remote, &ref, error))
-    return FALSE;
-  if (remote == NULL)
+  /* Handle the ref from the commit's metadata */
+  if (ref_for_deployment != NULL)
     {
-      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
-                   "Invalid refspec ‘%s’ in commit: did not contain a remote name",
-                   refspec);
-      return FALSE;
+      OstreeCollectionRef *orig_collection_ref = collection_ref;
+
+      /* Replace the ref, refspec and collection_ref */
+      g_free (ref);
+      ref = g_steal_pointer (&ref_for_deployment);
+      g_free (refspec);
+      refspec = g_strdup_printf ("%s:%s", remote, ref);
+      collection_ref = ostree_collection_ref_new (orig_collection_ref->collection_id,
+                                                  ref);
+      ostree_collection_ref_free (orig_collection_ref);
```

Comment:
Doing this free/replace dance has a strange smell to it, I have a suggestion for how this could be a little neater.

Reply:
Sure does. I couldn't decide a good scheme that let me keep part of the original call.